File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character,Integer> charMap = new HashMap<>();
        int left=0;
        int right=0;
        int len = 0;

        while (right<s.length()){
            char r = s.charAt(right);
            charMap.put(r, charMap.getOrDefault(r, 0)+1);
            
            while(charMap.get(r)>1){
                char l = s.charAt(left);
                charMap.put(l, charMap.get(l)-1);
                left++;
            }

            len = Math.max(len, right-left+1);

            right++;
        }

        return len;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res = 0
        left = 0
        right = 0
        chars = Counter()
        while right < len(s):
            r = s[right]
            chars[r] += 1

            while chars[r] > 1:
                l = s[left]
                chars[l] -= 1
                left += 1

            res = max(res, right - left + 1)
            right += 1

        return res
